dpdk/vr_dpdk_ringdev.c

```diff
@@ -49,23 +57,93 @@ dpdk_ring_allocate(unsigned host_lcore_id, unsigned vif_idx, unsigned for_lcore_
     return ring;
 }
 
+/* Free the ring */
+static void
+dpdk_ring_free(unsigned host_lcore_id, struct rte_ring *ring)
+{
+    struct vr_dpdk_lcore *host_lcore = vr_dpdk.lcores[host_lcore_id];
+
+    host_lcore->lcore_free_rings[host_lcore->lcore_nb_free_rings++] =
+        ring;
+}
+
 /* Add the ring to the list of rings to push */
 void
 dpdk_ring_to_push_add(unsigned lcore_id, struct rte_ring *tx_ring,
-    struct vr_dpdk_tx_queue *tx_queue)
+    struct vr_dpdk_queue *tx_queue)
 {
     struct vr_dpdk_lcore *lcore = vr_dpdk.lcores[lcore_id];
-    struct vr_dpdk_ring_to_push *ring_to_push =
-        &lcore->lcore_rings_to_push[lcore->lcore_nb_rings_to_push++];
+    struct vr_dpdk_ring_to_push *rtp = &lcore->lcore_rings_to_push[0];
 
-    RTE_VERIFY(lcore->lcore_nb_rings_to_push < VR_MAX_INTERFACES);
+    /* find an empty ring to push */
+    while (rtp->rtp_tx_ring) rtp++;
+
+    rtp->rtp_tx_ring = tx_ring;
+    rtp->rtp_tx_queue = tx_queue;
+    rte_wmb();
+    lcore->lcore_nb_rings_to_push++;
+    RTE_VERIFY(lcore->lcore_nb_rings_to_push < VR_DPDK_MAX_RINGS);
+}
 
-    ring_to_push->rtp_tx_ring = tx_ring;
-    ring_to_push->rtp_tx_queue = tx_queue;
+/* Remove the ring from the list of rings to push */
+void
+dpdk_ring_to_push_remove(unsigned lcore_id, struct rte_ring *tx_ring)
```

Comment:
This function will only be called from the netlink thread inside vrouter (i.e. there cannot be concurrent calls)?